docs/concepts/juttle_tutorial.md

```diff
@@ -2,27 +2,532 @@
+count the total number of points, the number of unique repository names, and the
+number of unique actors (in the "actor_login" field) like this:
+
+```
+read file -file '/tmp/data.json'
+| reduce total = count(), 
+    actors = count_unique('actor_login'), 
+    repos = count_unique('repo_name')
+```
+
+## Enriching the data (Put vs reduce)
+
+Let's say we wanted to know the number of actors participating in an average GitHub repository. The program above gives us the counts of actors and repos, so finding the average is simple math. However, it will not work to add "avg_actors = actors / repos" to the `reduce` expression, since this computation is not a reducer; and in general, this is not the logic you seek.
+
+If your intention is to add the computed field to each data point, thereby enriching and widening the dataset rather than reducing it, then it's not `reduce` you should be using but [`put`](../processors/put.md) processor.
+
+Let's compute the average actors per repo with an appended `put`.
+
+```
+read file -file '/tmp/data.json'
+| reduce total = count(), 
+    actors = count_unique('actor_login'), 
+    repos = count_unique('repo_name')
+| put avg_actors = actors / repos
+```
+
+In this program, `reduce` already turned our whole data set into a single data point, so `put` appended the new field to that one point. If there were many data points, `put` would have added the field to each one. We will try that out in the next section.
+
+## How many actors per repo? (Reduce with by-grouping)
+
+Unsurprisingly, our sample data set has many more actors than repositories. Maybe we'd like to know how many unique actors there are _per repository_. Easy! We just need to add a `by` clause to our reduce expression, indicating that we want to group data points by repository and carry out the `count_unique` operation individually for each group:
+
+```
+read file -file '/tmp/data.json'
+| reduce actors = count_unique('actor_login') by repo_name
+```
+
+If you want to see the top 10 repositories by number of participating actors, append the [`sort`](../processors/sort.md) processor with -desc order, then limit to first 10 points with the [`head`](../processors/head.md) processor. 
+
+```
+read file -file '/tmp/data.json'
+| reduce actors = count_unique('actor_login') by repo_name
+| sort actors -desc
+| head 10
+```
+
+Now for these most active repositories by participant count, let's see how many events per actor tend to occur. 
+
+```
+read file -file '/tmp/data.json'
+| reduce events = count(), actors = count_unique('actor_login') by repo_name
+| sort actors -desc
+| head 10
+| put activity_per_actor = Math.round(events/actors) by repo_name
+```
+
+This example clearly shows that `put` adds the newly computed field to each data point. Also, here we are using `put` with by-grouping (to be fair, the program above would run identically even with plain `put`, but it helps express the intention of the code).
```

Comment:
This parenthetical throws me off, let's replace it with a few words about what by-grouping does in general without admitting that it's not doing anything in this program.